File: src/test/java/com/fml/common/observability/FmlTelemetryTests.java
package com.fml.common.observability;

import static org.assertj.core.api.Assertions.assertThat;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.metrics.SdkMeterProvider;
import io.opentelemetry.sdk.metrics.data.MetricData;
import io.opentelemetry.sdk.testing.exporter.InMemoryMetricReader;
import java.util.List;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class FmlTelemetryTests {

    private static InMemoryMetricReader reader;

    @BeforeAll
    static void registerSdk() {
        GlobalOpenTelemetry.resetForTest();
        reader = InMemoryMetricReader.create();
        OpenTelemetrySdk sdk = OpenTelemetrySdk.builder()
                .setMeterProvider(SdkMeterProvider.builder().registerMetricReader(reader).build())
                .build();
        GlobalOpenTelemetry.set(sdk);
    }

    @Test
    void instrumentsKeepTheirNames() {
        assertThat(FmlTelemetry.SERVICE_NAME).isEqualTo("fml-monolith");
        assertThat(FmlTelemetry.TELEMETRY_READINGS_INGESTED).isEqualTo("fml.telemetry.readings_ingested");
        assertThat(FmlTelemetry.WORK_ORDERS_CREATED).isEqualTo("fml.maintenance.work_orders_created");
        assertThat(FmlTelemetry.PARTS_REORDERED).isEqualTo("fml.inventory.parts_reordered");
        assertThat(FmlTelemetry.REPORTS_GENERATED).isEqualTo("fml.reporting.reports_generated");
        assertThat(FmlTelemetry.tracer()).isNotNull();
        assertThat(FmlTelemetry.meter()).isNotNull();
    }

    @Test
    void counterMeasurementsAreCollected() {
        FmlTelemetry.workOrdersCreated().add(3);
        FmlTelemetry.telemetryReadingsIngested().add(1);
        FmlTelemetry.partsReordered().add(1);
        FmlTelemetry.reportsGenerated().add(1);

        List<MetricData> metrics = List.copyOf(reader.collectAllMetrics());

        assertThat(metrics)
                .extracting(MetricData::getName)
                .contains(
                        FmlTelemetry.TELEMETRY_READINGS_INGESTED,
                        FmlTelemetry.WORK_ORDERS_CREATED,
                        FmlTelemetry.PARTS_REORDERED,
                        FmlTelemetry.REPORTS_GENERATED);
        assertThat(metrics)
                .allSatisfy(metric ->
                        assertThat(metric.getInstrumentationScopeInfo().getName()).isEqualTo(FmlTelemetry.SERVICE_NAME));

        MetricData workOrders = metrics.stream()
                .filter(metric -> metric.getName().equals(FmlTelemetry.WORK_ORDERS_CREATED))
                .findFirst()
                .orElseThrow();
        assertThat(workOrders.getLongSumData().getPoints())
                .singleElement()
                .satisfies(point -> assertThat(point.getValue()).isEqualTo(3));
    }
}

File: tests/Fml.Common.Tests/FmlTelemetryTests.cs
using System.Diagnostics.Metrics;
using FML.Common.Observability;

namespace Fml.Common.Tests;

public class FmlTelemetryTests
{
    [Fact]
    public void Instruments_share_the_service_meter_and_keep_their_names()
    {
        Assert.Equal("fml-monolith", FmlTelemetry.ServiceName);
        Assert.Equal(FmlTelemetry.ServiceName, FmlTelemetry.ActivitySource.Name);

        var instruments = new Instrument[]
        {
            FmlTelemetry.TelemetryReadingsIngested,
            FmlTelemetry.WorkOrdersCreated,
            FmlTelemetry.PartsReordered,
            FmlTelemetry.ReportsGenerated,
        };

        Assert.All(instruments, instrument => Assert.Same(FmlTelemetry.Meter, instrument.Meter));
        Assert.Equal(
            new[]
            {
                "fml.telemetry.readings_ingested",
                "fml.maintenance.work_orders_created",
                "fml.inventory.parts_reordered",
                "fml.reporting.reports_generated",
            },
            instruments.Select(i => i.Name));
    }

    [Fact]
    public void Counter_measurements_are_observable_by_a_listener()
    {
        var counter = FmlTelemetry.WorkOrdersCreated;
        long total = 0;
        using var listener = new MeterListener();
        listener.InstrumentPublished = (instrument, l) =>
        {
            if (ReferenceEquals(instrument, counter))
            {
                l.EnableMeasurementEvents(instrument);
            }
        };
        listener.SetMeasurementEventCallback<long>((_, measurement, _, _) => total += measurement);
        listener.Start();

        counter.Add(3);

        Assert.Equal(3, total);
    }
}
